File: cplusplus/ext3/WorkerExt3.cpp
#include "WorkerExt3.hpp"

WorkerExt3::WorkerExt3(CoffeeHouse* model, CashDesk* cashDesk) : Worker(model, cashDesk) {
}

void WorkerExt3::prepareProductProcess(long currentTime) {
	if (currentRecipe != nullptr) {
		bool isFail = rand() % 2;
		if (isFail) {
			if (cashDesk->getFirstVisitor() == nullptr) {
				return;
			}
			cashDesk->getFirstVisitor()->failDesiredProduct(currentTime, currentProduct);
			prepareProductFinished(currentTime);
			return;
		}
	}
	Worker::prepareProductProcess(currentTime);
}

File: python/extension3/WorkerExt3.py
from __future__ import annotations
import random
from typing import TYPE_CHECKING

from python.root.Worker import Worker

if TYPE_CHECKING:
    from python.root.CashDesk import CashDesk
    from python.root.CoffeeHouse import CoffeeHouse


class WorkerExt3(Worker):
    def __init__(self, model: CoffeeHouse, cashDesk: CashDesk):
        super().__init__(model, cashDesk)

    def _prepareProductProcess(self, currentTime: int) -> None:
        if self._currentRecipe is not None:
            isFail = random.choice([True, False])
            if isFail:
                if self._cashDesk.getFirstVisitor() is None:
                    return None
                self._cashDesk.getFirstVisitor().failDesiredProduct(currentTime, self._currentProduct)
                self._prepareProductFinished(currentTime)
                return None
        super()._prepareProductProcess(currentTime)
